### ISSUE TITLE
Error finding Cfitsio library when building

### ISSUE THREAD

**Issue by VeerioSDSC** [healpy/healpy#954]
# Building with official source release

On a Linux OS, building Healpy has issues finding the cfitsio library. Binary installation with pip works, but I need to build from the official source release (working on a summer project with @rpwagner). 

Ran the following on a linux terminal (with Python 3.8):

```
curl -OL https://pypi.python.org/packages/source/h/healpy/healpy-1.14.0.tar.gz
tar -xzf healpy-1.14.0.tar.gz
cd healpy-1.14.0.tar.gz
pip install .
```

Output:
`configure: error: could not find the cfitsio library
  Traceback (most recent call last):
	File "/tmp/pip-req-build-2tztgih_/setup.py", line 165, in build_library
  	build_args = self.pkgconfig(pkg_config_name)
	File "/tmp/pip-req-build-2tztgih_/setup.py", line 130, in pkgconfig
  	token[index:].decode() for token in check_output(cmd, env=env).split()
	File "/usr/lib/python3.8/subprocess.py", line 415, in check_output
  	return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
	File "/usr/lib/python3.8/subprocess.py", line 516, in run
  	raise CalledProcessError(retcode, process.args,
  subprocess.CalledProcessError: Command '('pkg-config', '--cflags-only-I', 'healpix_cxx >= 3.60.0')' returned non-zero exit status 1.
`
# Building from cloning from GitHub
Similar issue with cfitsio library happened after cloning from GitHub (Also happened on MacOS)

```
git clone https://github.com/healpy/healpy
cd healpy
python3 setup.py build
```

Output:
```

/usr/lib/python3.8/distutils/extension.py:131: UserWarning: Unknown Extension options: 'cython_directives'
  warnings.warn(msg)
running build
running build_clib
checking if configure script for library 'cfitsio' exists
running 'autoreconf -i' for library 'cfitsio'
autoreconf: 'configure.ac' or 'configure.in' is required
```

Please let me know what I should try next or if you have any questions. Thank you

**Comment by zonca** [comment]
On Mon, Jul 8, 2024, 11:35 VeerioSDSC ***@***.***> wrote:

> Building with official source release
>
> On a Linux OS, building Healpy has issues finding the cfitsio library.
> Binary installation with pip works, but I need to build from the official
> source release (working on a summer project with @rpwagner
> <https://github.com/rpwagner>).
>
> Ran the following on a linux terminal (with Python 3.8):
>
> curl -OL https://pypi.python.org/packages/source/h/healpy/healpy-1.14.0.tar.gz
> tar -xzf healpy-1.14.0.tar.gz
> cd healpy-1.14.0.tar.gz
> pip install .
>
> Output:
> configure: error: could not find the cfitsio library Traceback (most
> recent call last): File "/tmp/pip-req-build-2tztgih_/setup.py", line 165,
> in build_library build_args = self.pkgconfig(pkg_config_name) File
> "/tmp/pip-req-build-2tztgih_/setup.py", line 130, in pkgconfig
> token[index:].decode() for token in check_output(cmd, env=env).split() File
> "/usr/lib/python3.8/subprocess.py", line 415, in check_output return
> run(*popenargs, stdout=PIPE, timeout=timeout, check=True, File
> "/usr/lib/python3.8/subprocess.py", line 516, in run raise
> CalledProcessError(retcode, process.args, subprocess.CalledProcessError:
> Command '('pkg-config', '--cflags-only-I', 'healpix_cxx >= 3.60.0')'
> returned non-zero exit status 1.
>

This is a very old version of healthy, try with 1.17.1


Building from cloning from GitHub
>
> Similar issue with cfitsio library happened after cloning from GitHub
> (Also happened on MacOS)
>
> git clone https://github.com/healpy/healpy
> cd healpy
> python3 setup.py build
>
> Output:
>
>
> /usr/lib/python3.8/distutils/extension.py:131: UserWarning: Unknown Extension options: 'cython_directives'
>   warnings.warn(msg)
> running build
> running build_clib
> checking if configure script for library 'cfitsio' exists
> running 'autoreconf -i' for library 'cfitsio'
> autoreconf: 'configure.ac' or 'configure.in' is required
>
>
Have you run

git submodule init
git submodule update

To grab the submodules?

**Comment by VeerioSDSC** [comment]
Thank you for the help. Both methods worked to build healpy normally!

However, when I try to build with [Pyodide](https://github.com/pyodide/pyodide.git) I run into the same error when building with healpy. 

Ran the following: 
```
curl -OL https://pypi.python.org/packages/source/h/healpy/healpy-1.17.1.tar.gz
tar -xvf healpy-1.17.1.tar.gz         
cd healpy-1.17.1
git submodule init
git submodule update
pyodide build
```

Which gave me a similar output:
```
configure: error: could not find the cfitsio library
Traceback (most recent call last):
  File "<string>", line 126, in build_library
  File "<string>", line 91, in pkgconfig
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/subprocess.py", line 466, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '('pkg-config', '--cflags-only-I', 'healpix_cxx >= 3.83.0')' returned non-zero exit status 1.

```

Please let me know if you have any suggestions on how to fix this issue when building with Pyodide. Thank you

**Comment by rpwagner** [comment]
@VeerioSDSC were you trying on Linux for the last attempt?

**Comment by VeerioSDSC** [comment]
Hello, 

Thank you - I have tried it on linux now, but I am now running into this error when building from GitHub.

Commands I ran:

```
git clone https://github.com/healpy/healpy
cd healpy
git submodule init
git submodule update
python3 setup.py build
```
Output:
```
checking if configure script for library 'sharp' exists
running 'autoreconf -i' for library 'sharp'
Traceback (most recent call last):
```
 
(lot of traceback lines) which ends up to this:
```
 File "/home/veer/anaconda3/lib/python3.12/subprocess.py", line 1872, in _execute_child
    env_list.append(k + b'=' + os.fsencode(v))
                               ^^^^^^^^^^^^^^
  File "<frozen os>", line 847, in fsencode
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

Please let me know if you have any suggestions for this. Huge thanks.

**Comment by zonca** [comment]
are you following https://github.com/healpy/healpy/blob/main/INSTALL.rst?

**Comment by VeerioSDSC** [comment]
Thank you.

However, after running` pip install . ` with healpy1.17.1, I get:

```
configure: error: C compiler cannot create executables
      See `config.log' for more details
      Traceback (most recent call last):
```

which leads to:
```

File "/home/veer/anaconda3/lib/python3.12/subprocess.py", line 413, in check_call
          raise CalledProcessError(retcode, cmd)
      subprocess.CalledProcessError: Command '['/bin/sh', '/home/veer/pyodideoutoftree/healpy-1.17.1/cextern/healpix/src/common_libraries/libsharp/configure', '--prefix=/home/veer/pyodideoutoftree/healpy-1.17.1/build/temp.linux-x86_64-cpython-312', '--disable-curl', '--with-pic', '--disable-maintainer-mode']' returned non-zero exit status 77.
```

My config.log file is 
```

am__fastdepCC_TRUE=''
am__include='include'
am__isrc=' -I$(srcdir)'
am__leading_dot='.'
am__nodep='_no'
am__quote=''
am__tar='$${TAR-tar} chof - "$$tardir"'
am__untar='$${TAR-tar} xf -'
bindir='${exec_prefix}/bin'
build=''
build_alias=''
build_cpu=''
build_os=''
build_vendor=''
datadir='${datarootdir}'
datarootdir='${prefix}/share'
docdir='${datarootdir}/doc/${PACKAGE_TARNAME}'
dvidir='${docdir}'
exec_prefix='NONE'
host=''
host_alias=''
host_cpu=''
host_os=''
host_vendor=''
htmldir='${docdir}'
includedir='${prefix}/include'
infodir='${datarootdir}/info'
install_sh='${SHELL} /home/veer/pyodideoutoftree/healpy-1.17.1/cextern/healpix/src/common_libraries/libsharp/install-sh'
libdir='${exec_prefix}/lib'
libexecdir='${exec_prefix}/libexec'
localedir='${datarootdir}/locale'
localstatedir='${prefix}/var'
mandir='${datarootdir}/man'
mkdir_p='$(MKDIR_P)'
oldincludedir='/usr/include'
pdfdir='${docdir}'
pkgconfig_libdir=''
pkgconfig_libfile=''
prefix='/home/veer/pyodideoutoftree/healpy-1.17.1/build/temp.linux-x86_64-cpython-312'
program_transform_name='s,x,x,'
psdir='${docdir}'
runstatedir='${localstatedir}/run'
sbindir='${exec_prefix}/sbin'
sharedstatedir='${prefix}/com'
sysconfdir='${prefix}/etc'
target_alias=''

## ----------- ##
## confdefs.h. ##
## ----------- ##

/* confdefs.h */
#define PACKAGE_NAME "libsharp"
#define PACKAGE_TARNAME "libsharp"
#define PACKAGE_VERSION "1.0.0"
#define PACKAGE_STRING "libsharp 1.0.0"
#define PACKAGE_BUGREPORT ""
#define PACKAGE_URL ""
#define PACKAGE "libsharp"
#define VERSION "1.0.0"
```

Thank you for your continued help.

**Comment by VeerioSDSC** [comment]
Update: Fixed previous issue. Please disregard last comment.

However, still running into error finding cfitsio library when building, following instructions from INSTALL.rst as follows - using the most recent version of healpy 1.17.1:

```
curl -OL https://pypi.python.org/packages/source/h/healpy/healpy-1.17.1.tar.gz
tar -xzf healpy-1.17.1.tar.gz
cd healpy-1.17.1/
pip install .
```
```
configure: error: could not find the cfitsio library
Traceback (most recent call last):
```

(some traceback lines)
```
 subprocess.CalledProcessError: Command '['/bin/sh', '/home/veer/pyodideoutoftree/healpy-1.17.1/cextern/healpix/src/cxx/configure', '--prefix=/home/veer/pyodideoutoftree/healpy-1.17.1/build/temp.linux-x86_64-cpython-312', '--disable-curl', '--with-pic', '--disable-maintainer-mode']' returned non-zero exit status 1.
   
```

Not cloning from GitHub, so cannot use git submodule init/git submodule update. 

Please let me know, thanks

**Comment by rpwagner** [comment]
You should keep trying with the building within the git repo. The healpy source alone from the tarball doesn’t include the needed libraries like cfitsio. In the end, you want a build process that also compiles those libraries to WebAssembly so Pyodide can use them.

**Comment by VeerioSDSC** [comment]
Thank you. 

I have been running into a similar problem when [building with Pyodide](https://pyodide.org/en/stable/development/building-and-testing-packages.html) from GitHub:

After running `pyodide build` I get an output issue of:

```
configure: error: Unable to locate zlib compression library
Traceback (most recent call last):
  File "<string>", line 126, in build_library
  File "<string>", line 91, in pkgconfig
```
leading to 
```
File "/home/veer/anaconda3/lib/python3.12/subprocess.py", line 413, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/bin/sh', '/home/veer/pyodideoutoftree/healpy/cextern/cfitsio/configure', '--prefix=/home/veer/pyodideoutoftree/healpy/build/temp.emscripten_3_1_58_wasm32-cpython-312', '--disable-curl', '--with-pic', '--disable-maintainer-mode']' returned non-zero exit status 1.

```

So far, I have tried building with other packages with pyodide, which has worked. 

I also have tried fixing with the development package with `sudo apt-get install zlib1g` and `sudo apt-get install zlib1g-dev`, still no luck. 

I also have tried editing the PKG_CONFIG environment variable to /usr/bin/pkg-config, but still no luck. 

Thank you for your continued help :)

**Comment by VeerioSDSC** [comment]
Thank you. 

I have been running into the same earlier problem when doing [development install ](https://healpy.readthedocs.io/en/latest/install.html#:~:text=Development%20install,the%20pip%20section.) on Ubuntu. 

Ran the following:

```
git clone https://github.com/healpy/healpy.git
cd healpy
sudo apt-get install autoconf automake libtool pkg-config
git submodule init
git submodule update
sudo apt-get install libcfitsio-dev
sudo apt-get install libsharp-dev
pip install -e .
```
(had to install the dev packages as said packages not found initially)

 ```
configure: error: could not find the cfitsio library
      Traceback (most recent call last):
        File "<string>", line 126, in build_library
        File "<string>", line 91, in pkgconfig
```

Also tried building it with the flags: `CC=gcc CXX=g++ CFLAGS=’-fPIC’ CXXFLAGS=’-fPIC’ pip install -e .` but realized it wasn't a compiler issue. 

Still had the same issue

Thank you for your continued help :)

**Comment by rpwagner** [comment]
@zonca, FYI I’ve had the same issues as @VeerioSDSC building on a fresh Ubuntu host. The healpix build is having trouble finding libraries.

**Comment by zonca** [comment]
I'll look into the ubuntu issue later on, for testing with piodide, let's switch to https://github.com/healpy/pyhealpy/issues/1

**Comment by VeerioSDSC** [comment]
Thank you for your help, I have sent it in [here](https://github.com/healpy/pyhealpy/issues/1#issuecomment-2222080988)

**Comment by badshah400** [comment]
I don't know if this is the same problem we see when building healpy 1.17.3 on openSUSE Tumbleweed, but the error log seems to be at least similar:

```log
[   25s]   checking pkg-config is at least version 0.9.0... yes
[   25s]   checking for cfitsio... yes
[   25s]   checking for ffgnrwll in -lcfitsio... no
[   25s]   configure: error: could not find the cfitsio library
[   25s]   Traceback (most recent call last):
[   25s]     File "<string>", line 126, in build_library
[   25s]     File "<string>", line 91, in pkgconfig
[   25s]     File "/usr/lib64/python3.10/subprocess.py", line 421, in check_output
[   25s]       return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
[   25s]     File "/usr/lib64/python3.10/subprocess.py", line 526, in run
[   25s]       raise CalledProcessError(retcode, process.args,
[   25s]   subprocess.CalledProcessError: Command '('pkg-config', '--cflags-only-I', 'healpix_cxx >= 3.83.0')' returned non-zero exit status 1.
[   25s] 
[   25s]   During handling of the above exception, another exception occurred:
[   25s] 
[   25s]   Traceback (most recent call last):
[   25s]     File "/usr/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
[   25s]       main()
[   25s]     File "/usr/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
[   25s]       json_out['return_val'] = hook(**hook_input['kwargs'])
[   25s]     File "/usr/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 251, in build_wheel
[   25s]       return _build_backend().build_wheel(wheel_directory, config_settings,
[   25s]     File "/usr/lib/python3.10/site-packages/setuptools/build_meta.py", line 415, in build_wheel
[   25s]       return self._build_with_temp_dir(
[   25s]     File "/usr/lib/python3.10/site-packages/setuptools/build_meta.py", line 397, in _build_with_temp_dir
[   25s]       self.run_setup()
[   25s]     File "/usr/lib/python3.10/site-packages/setuptools/build_meta.py", line 313, in run_setup
[   25s]       exec(code, locals())
[   25s]     File "<string>", line 304, in <module>
[   25s]     File "/usr/lib/python3.10/site-packages/setuptools/__init__.py", line 103, in setup
[   25s]       return distutils.core.setup(**attrs)
[   25s]     File "/usr/lib/python3.10/site-packages/setuptools/_distutils/core.py", line 184, in setup
[   25s]       return run_commands(dist)
[   25s]     File "/usr/lib/python3.10/site-packages/setuptools/_distutils/core.py", line 200, in run_commands
[   25s]       dist.run_commands()
[   25s]     File "/usr/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 969, in run_commands
[   25s]       self.run_command(cmd)
[   25s]     File "/usr/lib/python3.10/site-packages/setuptools/dist.py", line 976, in run_command
[   25s]       super().run_command(command)
[   25s]     File "/usr/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
[   25s]       cmd_obj.run()
[   25s]     File "/usr/lib/python3.10/site-packages/setuptools/command/bdist_wheel.py", line 373, in run
[   25s]       self.run_command("build")
[   25s]     File "/usr/lib/python3.10/site-packages/setuptools/_distutils/cmd.py", line 316, in run_command
[   25s]       self.distribution.run_command(command)
[   25s]     File "/usr/lib/python3.10/site-packages/setuptools/dist.py", line 976, in run_command
[   25s]       super().run_command(command)
[   25s]     File "/usr/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
[   25s]       cmd_obj.run()
[   25s]     File "/usr/lib/python3.10/site-packages/setuptools/_distutils/command/build.py", line 132, in run
[   25s]       self.run_command(cmd_name)
[   25s]     File "/usr/lib/python3.10/site-packages/setuptools/_distutils/cmd.py", line 316, in run_command
[   25s]       self.distribution.run_command(command)
[   25s]     File "/usr/lib/python3.10/site-packages/setuptools/dist.py", line 976, in run_command
[   25s]       super().run_command(command)
[   25s]     File "/usr/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
[   25s]       cmd_obj.run()
[   25s]     File "/usr/lib/python3.10/site-packages/setuptools/_distutils/command/build_clib.py", line 112, in run
[   25s]       self.build_libraries(self.libraries)
[   25s]     File "<string>", line 256, in build_libraries
[   25s]     File "<string>", line 178, in build_library
[   25s]     File "/usr/lib64/python3.10/subprocess.py", line 369, in check_call
[   25s]       raise CalledProcessError(retcode, cmd)
[   25s]   subprocess.CalledProcessError: Command '['/bin/sh', '/home/abuild/rpmbuild/BUILD/healpy-1.17.1/cextern/healpix/src/cxx/configure', '--prefix=/home/abuild/rpmbuild/BUILD/healpy-1.17.1/build/temp.linux-x86_64-cpython-310', '--disable-curl', '--with-pic', '--disable-maintainer-mode']' returned non-zero exit status 1.
[   25s]   error: subprocess-exited-with-error
[   25s]   
[   25s]   × Building wheel for healpy (pyproject.toml) did not run successfully.
[   25s]   │ exit code: 1
[   25s]   ╰─> See above for output.
```

This is because we try to build against system installed `cfitsio` shared library, while the bundled `healpix_cxx` sources in [./cextern](../tree/main/cextern) uses `PKG_CHECK_MODULES_STATIC` to look for `cfitsio`, leading to linking failures and therefore the inability to find `ffgnrwll`. If this is the same issue, you will need, as we do, [r1240](https://sourceforge.net/p/healpix/code/1240/) from healpix upstream, i.e., the following patch:

```diff
--- a/trunk/src/cxx/configure.ac
+++ b/trunk/src/cxx/configure.ac
@@ -50,11 +50,7 @@
 dnl Now check for cfitsio using pkg-config. If cfitsio is not found with
 dnl pkg-config, CFITSIO_LIBS defaults to "-lcfitsio".
 dnl
-dnl FIXME: Use PKG_CHECK_MODULES_STATIC to pick up additional dependencies
-dnl until cfitsio gains a modernized automake/libtool build system.
-dnl See https://github.com/HEASARC/cfitsio/issues/18,
-dnl https://github.com/healpy/healpy/pull/917.
-PKG_CHECK_MODULES_STATIC([CFITSIO], [cfitsio], , [CFITSIO_LIBS=-lcfitsio])
+PKG_CHECK_MODULES([CFITSIO], [cfitsio], , [CFITSIO_LIBS=-lcfitsio])
 dnl
 dnl With pkg-config < 0.24, PKG_CHECK_MODULES does not AC_SUBST the _CFLAGS
 dnl and _LIBS variables.
```

**Comment by lpsinger** [comment]
> This is because we try to build against system installed `cfitsio` shared library, while the bundled `healpix_cxx` sources in [./cextern](../tree/main/cextern) uses `PKG_CHECK_MODULES_STATIC` to look for `cfitsio`, leading to linking failures and therefore the inability to find `ffgnrwll`.

Thee comments in the diff above contain some pertinent links:

> FIXME: Use PKG_CHECK_MODULES_STATIC to pick up additional dependencies
> until cfitsio gains a modernized automake/libtool build system.
> See https://github.com/HEASARC/cfitsio/issues/18,
> https://github.com/healpy/healpy/pull/917.

The next release of CFITSIO will have some build system changes that make the `PKG_CHECK_MODULES_STATIC` hack unnecessary. Sorry that this caused you inconvenience! We are aware of this and working with the CFITSIO developers to improve matters.

**Comment by VeerioSDSC** [comment]
Thank you for letting us know :)

**Comment by lpsinger** [comment]
Fixed by #942 (although not yet released).
